Multiple New Session Recording Features

GPayne opened the pull request:
Resolves issues #1373 and #1506. Mostly resolves issue#1347, and some other requests that never made it to being issues:

- Fixed-framerate playback for HD rendering playback
- Time rate at beginning of record is preserved
- For every property change made during a recording, a corresponding property command is added to the start of the file to preserve its initial conditions
- A recorded file is parsed before playback starts, and is rejected if it contains asset(s) that are not currently loaded
- Pausing during playback is now supported
- A reject list will prevent certain scripts from being recorded (which might interfere with playback or with editing playback files)
- Now records to timeline in memory, saving to disk when recording has stopped

alexanderbock commented:
Found a bug:
1. Set delta time 50 some value 5 minute per second
2. Pause delta time
3. Start recording, move camera, and unpause while recording is playing
4. When playing back the recording, the camera is pointing in the wrong direction

I also managed to somehow managed to make it such that the delta time is 0 even though it is just paused but has a deltatime != 0

alexanderbock commented:
As suspected during the call:
1. Set the delta time to something != 0 and 1
2. Pause the time
3. Start recoding
->  The script will say `script 96.2297 0 673298785.483  1 openspace.time.setDeltaTime(0)`  I don't think it will be an issue with the recording as the sim time is set anyway, but it might be confusing if someone looks at the script afterwards

alexanderbock commented:
Managed to produce a crash:
1. Start a recording
2. Add a screenspace image through the screenspace renderables UI (`openspace.addScreenSpaceRenderable({["Identifier"] = "asd",["Name"] = "asd",["Type"] = "ScreenSpaceImageOnline",["URL"] = "https://static.remove.bg/remove-bg-web/2a274ebbb5879d870a69caae33d94388a88e0e35/assets/start-0e837dcc57769db2306d8d659f53555feb500b3c5d456879b9c843d1872e7baa.jpg"})`)
3. Toggling the visibility of the screenspace renderable

This causes a crash in:
sessionrecording.cpp:1445

alexanderbock commented:
I didn't manage to find the script used to pause the playback

WeirdRubberDuck commented on src/interaction/sessionrecording.cpp:
```diff
@@ -1326,83 +1599,114 @@ bool SessionRecording::readScriptKeyframeAscii(Timestamps& times,
     double currTime = currentTime();
     lookForNonCameraKeyframesThatHaveComeDue(currTime);
     updateCameraWithOrWithoutNewKeyframes(currTime);
+    //Unfortunately the first frame is sometimes rendered because globebrowsing reports
```
incredibly nitpicky, but I would add a whitespace before "Unfortunately" and remove the dot at the end of this comment 😄

WeirdRubberDuck reviewed (commented):
Also managed to find a bug related to pausing the simulation/delta time, but maybe it is the same as Alex found: 

1. Make a recording where the delta time **is not** paused (either for a part of or for the entire recording)
2. After finished recording, pause the simulation time
3. Play back the created recording
4. The camera will point in the wrong direction (E.g. I did a recording around Earth and when playing it back Earth was no longer visible)

Alternatively

1. Start recording where the delta time **is** paused. Keep it paused during the entire recording
2. Move the camera
3. Stop recording
4. Playback the recording with the delta time being paused
5. The camera will not move

WeirdRubberDuck commented:
Also, it seems like pausing the simulation time/delta time is not correctly recorded:

1. Start a recording with simulation delta time !=0 and not paused
2. Pause the delta time while recording
3. Stop recording
4. Set delta time to not be paused again
5. Start the created recording. The simulation time will not be paused during the recording

Thinking that maybe I have misunderstood something about how the "pause" is supposed to affect the session recordings? 🤔 My impression is that it should be recorded, just as any other changes to the simulation time

GPayne commented:
Implemented a playback-pause which is separate from pausing the delta time. A GUI element needs to be added in order to show this playback-pause state to user.

alexanderbock commented:
UI component for exporting frames during playback with frames/s counter

GPayne commented on include/openspace/interaction/sessionrecording.h:
```diff
@@ -616,27 +659,65 @@ class SessionRecording : public properties::PropertyOwner {
+        "NavigationHandler.OrbitalNavigator.RetargetAnchor",
+        "NavigationHandler.OrbitalNavigator.RetargetAim"
+    };
+    //A script that begins with an exact match of any of the strings contained in _scriptRejects will not be recorded
```
Fixed

GPayne commented on src/interaction/sessionrecording.cpp:
```diff
@@ -640,27 +766,37 @@ void SessionRecording::saveTimeKeyframeAscii(Timestamps& times,
-    saveSingleKeyframeScript(
-        sm,
+void SessionRecording::saveScriptKeyframeToPropertiesBaseline(std::string script)
+{
```
Done

GPayne commented on src/interaction/sessionrecording.cpp:
```diff
@@ -689,13 +825,45 @@ void SessionRecording::saveScriptKeyframeAscii(Timestamps& times,
+bool SessionRecording::isPropertyAllowedForBaseline(const std::string& propString) {
+    for (std::string reject : _propertyBaselineRejects) {
+        if (propString.substr(0, reject.length()).compare(reject) == 0)
+        {
```
Did

GPayne commented on src/interaction/sessionrecording.cpp:
```diff
@@ -1015,7 +1182,7 @@ bool SessionRecording::readCameraKeyframeBinary(Timestamps& times,
         return false;
     }
-    times.timeOs = kf._timestamp;
+    //times.timeOs = kf._timestamp;
```
Removed leftovers

GPayne commented on include/openspace/interaction/sessionrecording.h:
```diff
@@ -616,27 +659,65 @@ class SessionRecording : public properties::PropertyOwner {
+        "openspace.time.togglePause"
+    };
+    const std::vector<std::string> _navScriptsUsingNodes = {
+    "RetargetAnchor",
```
Good find! I just repeated myself while trying to implement this change.

GPayne commented on src/interaction/sessionrecording.cpp:
```diff
@@ -223,25 +228,117 @@ bool SessionRecording::startRecording(const std::string& filename) {
+        }
+        for (timelineEntry entry : _timeline) {
+            switch (entry.keyframeType) {
+            case RecordedType::Camera:
```
This is one of those cases (pun intended) where I've seen it done both ways. I agree that it's more readable indented, but this way seems to be more common (K&R style, java style, etc). I can't say I've been consistent with this because previously I wrote an indented case in the same file. I'm going to stick with this way here, particularly because it's already indented a lot and I won't have to wrap any lines :)

GPayne commented:
Commit 413639e fixes the bugs reported above, which came down to 2 things:

1. In the previous version, the pause state wasn't recorded
2. Problem with the screenspace renderable was fixed

GPayne commented:
This branch has some new changes since the last review:
* Pause within playback now works
* Loop playback mode was added
* Merge with master to get up to date

Went through full testing document (see the internal wiki) to check for any breaking changes.

WeirdRubberDuck commented on src/interaction/sessionrecording_lua.inl:
```diff
@@ -98,15 +112,15 @@ int startPlayback(lua_State* L, interaction::KeyframeTimeRef timeMode,
 }
 
 int startPlaybackDefault(lua_State* L) {
-    ghoul::lua::checkArgumentsAndThrow(L, 1, "lua::startPlaybackDefault");
```
Instead of removing this check (which is really useful as it provides meaningful error info to the user), you can specify that one to two arguments are allowed like this: 
```
ghoul::lua::checkArgumentsAndThrow(L, {1, 2}, "lua::startPlaybackDefault");
```

GPayne replied:
Done

WeirdRubberDuck commented on src/interaction/sessionrecording_lua.inl:
```diff
@@ -120,7 +134,6 @@ int startPlaybackApplicationTime(lua_State* L) {
 }
 
 int startPlaybackRecordedTime(lua_State* L) {
-    ghoul::lua::checkArgumentsAndThrow(L, 1, "lua::startPlaybackRecordedTime");
```
see above

WeirdRubberDuck commented:
Tried it out a bit with the new UI and it works great 😄 I think to "loop" feature is really useful as well
